File: cypher/core/reverb.py
"""Dattorro plate reverb — based on Jon Dattorro's 1997 paper.

Topology (figure-8):
    input → pre-delay → input diffusion (4 allpass) → tank
    tank = two cross-fed loops, each: modulated allpass → delay → damping → decay
    output = tapped from multiple points in the tank

Parameters:
    mix       — dry/wet blend (0.0 = fully dry, 1.0 = fully wet)
    decay     — feedback amount in tank loops (0.0–1.0)
    damping   — high-frequency absorption in feedback (0.0 = bright, 1.0 = dark)
    predelay  — milliseconds before reverb onset
    size      — scales all delay lengths (room size)
    mod_depth — LFO modulation depth on tank allpasses
    mod_rate  — LFO modulation rate in Hz

C++ portability notes:
    - All delay lines are fixed-size arrays with power-of-2 masking.
    - Allpass and damping filters are single-multiply operations.
    - Modulated reads use linear interpolation (one branch, two multiplies).
    - Total memory: ~45KB at 48kHz. Fits comfortably in L1 cache.
    - No dynamic allocation. All buffers sized at init.
"""

from __future__ import annotations

import math

import numpy as np

from .types import AudioBuffer, DEFAULT_SAMPLE_RATE

# ── Dattorro delay lengths (samples at 29761 Hz, scaled to actual SR) ──
# These are the original prime-number lengths from the paper.
# We scale them to the actual sample rate at init time.
_REF_SR = 29761

# Input diffusion allpass lengths
_INPUT_DIFF_LENGTHS = [142, 107, 379, 277]

# Tank allpass lengths (one per loop)
_TANK_AP_LENGTHS = [672, 908]

# Tank delay lengths (one per loop)
_TANK_DELAY_LENGTHS = [4453, 4217]

# Additional tank delays (after damping, before cross-feed)
_TANK_DELAY2_LENGTHS = [3720, 3163]

# Output tap positions (relative to tank delays)
_TAP_POSITIONS_L = [266, 2974, 1913, 1996, 1990, 187, 1066]
_TAP_POSITIONS_R = [353, 3627, 1228, 2673, 2111, 335, 121]


def _next_pow2(n: int) -> int:
    """Round up to the next power of 2."""
    p = 1
    while p < n:
        p <<= 1
    return p


def _scale_length(ref_len: int, sample_rate: int) -> int:
    """Scale a delay length from reference SR to actual SR."""
    return max(1, int(round(ref_len * sample_rate / _REF_SR)))


class DelayLine:
    """Fixed-size circular buffer with fractional read support.

    C++ equivalent: float buffer[N] with a write index and mask.
    """

    __slots__ = ("_buf", "_mask", "_write_pos")

    def __init__(self, max_length: int) -> None:
        size = _next_pow2(max_length + 16)  # headroom for modulation
        self._buf = np.zeros(size, dtype=np.float32)
        self._mask = size - 1
        self._write_pos: int = 0

    def write(self, sample: float) -> None:
        self._buf[self._write_pos & self._mask] = sample
        self._write_pos += 1

    def read(self, delay: int) -> float:
        return float(self._buf[(self._write_pos - delay) & self._mask])

    def read_frac(self, delay: float) -> float:
        """Linear interpolation for fractional delay (modulated reads)."""
        d_int = int(delay)
        frac = delay - d_int
        a = float(self._buf[(self._write_pos - d_int) & self._mask])
        b = float(self._buf[(self._write_pos - d_int - 1) & self._mask])
        return a + frac * (b - a)

    def clear(self) -> None:
        self._buf.fill(0.0)
        self._write_pos = 0


class AllpassFilter:
    """Schroeder allpass: y[n] = -g*x[n] + x[n-d] + g*y[n-d].

    C++ equivalent: one delay line + one coefficient.
    """

    __slots__ = ("_delay", "_length", "_g")

    def __init__(self, length: int, gain: float = 0.5) -> None:
        self._delay = DelayLine(length)
        self._length = length
        self._g = gain

    def process(self, x: float) -> float:
        delayed = self._delay.read(self._length)
        v = x + self._g * delayed
        self._delay.write(v)
        return delayed - self._g * v

    def clear(self) -> None:
        self._delay.clear()


class ModulatedAllpass:
    """Allpass with LFO-modulated delay length for chorus-like smearing.

    The modulation breaks up metallic resonances — this is what gives
    Valhalla-style reverbs their lush, detuned character.
    """

    __slots__ = ("_delay", "_base_length", "_g")

    def __init__(self, length: int, gain: float = 0.5) -> None:
        self._delay = DelayLine(length + 64)  # extra room for mod excursion
        self._base_length = length
        self._g = gain

    def process(self, x: float, mod_offset: float = 0.0) -> float:
        read_pos = self._base_length + mod_offset
        read_pos = max(1.0, read_pos)
        delayed = self._delay.read_frac(read_pos)
        v = x + self._g * delayed
        self._delay.write(v)
        return delayed - self._g * v

    def clear(self) -> None:
        self._delay.clear()


class OnePoleLP:
    """One-pole lowpass for damping. y[n] = (1-g)*x[n] + g*y[n-1].

    g=0: no filtering (bright). g→1: heavy damping (dark).
    """

    __slots__ = ("_state",)

    def __init__(self) -> None:
        self._state: float = 0.0

    def process(self, x: float, damp: float) -> float:
        self._state = (1.0 - damp) * x + damp * self._state
        return self._state

    def clear(self) -> None:
        self._state = 0.0


class DattorroPlateReverb:
    """Dattorro plate reverb — stereo-out from mono input.

    Designed for use as a global send effect. Feed it the mixed mono
    signal, get back a stereo pair (or sum to mono for our setup).

    All delay lengths are scaled from Dattorro's reference rate (29761 Hz)
    to the actual sample rate, preserving the room character.
    """

    def __init__(self, sample_rate: int = DEFAULT_SAMPLE_RATE) -> None:
        self.sample_rate = sample_rate

        # ── Parameters (set directly or via Parameter objects) ──
        self.mix: float = 0.3          # dry/wet
        self.decay: float = 0.7        # tank feedback (0–0.99)
        self.damping: float = 0.3      # HF absorption (0=bright, 1=dark)
        self.predelay_ms: float = 20.0  # ms before reverb onset
        self.mod_depth: float = 0.5    # LFO excursion in samples
        self.mod_rate: float = 0.8     # LFO rate in Hz

        # ── Pre-delay ──
        max_predelay = int(0.15 * sample_rate)  # 150ms max
        self._predelay = DelayLine(max_predelay)

        # ── Input diffusion: 4 allpass stages ──
        self._input_aps = [
            AllpassFilter(_scale_length(l, sample_rate), g)
            for l, g in zip(
                _INPUT_DIFF_LENGTHS,
                [0.75, 0.75, 0.625, 0.625],
            )
        ]

        # ── Tank: two cross-fed loops ──
        # Each loop: modulated allpass → delay → damping → decay → delay2

        # Modulated allpasses
        self._tank_aps = [
            ModulatedAllpass(_scale_length(l, sample_rate), 0.7)
            for l in _TANK_AP_LENGTHS
        ]

        # Main tank delays
        self._tank_delays = [
            DelayLine(_scale_length(l, sample_rate))
            for l in _TANK_DELAY_LENGTHS
        ]
        self._tank_delay_lengths = [
            _scale_length(l, sample_rate) for l in _TANK_DELAY_LENGTHS
        ]

        # Secondary tank delays (after damping)
        self._tank_delays2 = [
            DelayLine(_scale_length(l, sample_rate))
            for l in _TANK_DELAY2_LENGTHS
        ]
        self._tank_delay2_lengths = [
            _scale_length(l, sample_rate) for l in _TANK_DELAY2_LENGTHS
        ]

        # Damping filters (one per loop)
        self._damp = [OnePoleLP(), OnePoleLP()]

        # Tank state (cross-fed between loops)
        self._tank_state = [0.0, 0.0]

        # ── LFO state ──
        self._lfo_phase: float = 0.0

        # ── Output tap positions (scaled) ──
        self._taps_l = [_scale_length(t, sample_rate) for t in _TAP_POSITIONS_L]
        self._taps_r = [_scale_length(t, sample_rate) for t in _TAP_POSITIONS_R]

    def process(self, input_buf: AudioBuffer) -> AudioBuffer:
        """Process a mono input buffer, return mono output (summed stereo).

        Hot path is fully inlined — no method calls, no attribute lookups
        in the inner loop. All delay lines accessed as raw numpy arrays
        with index masking. ~5-10x faster than the method-call version.
        """
        n = len(input_buf)
        output = np.zeros(n, dtype=np.float32)

        predelay_samps = max(1, int(self.predelay_ms / 1000.0 * self.sample_rate))
        decay = min(0.99, max(0.0, self.decay))
        damping = min(0.99, max(0.0, self.damping))
        damp_inv = 1.0 - damping
        mod_depth = self.mod_depth * 12.0
        lfo_inc = self.mod_rate * 2.0 * math.pi / self.sample_rate
        wet_gain = self.mix
        dry_gain = 1.0 - self.mix

        # --- Pull all arrays and state into locals (avoids self. lookups) ---
        pd_buf = self._predelay._buf
        pd_mask = self._predelay._mask
        pd_wp = self._predelay._write_pos

        # Input diffusion: 4 allpasses — extract raw buffers
        iap_bufs = [ap._delay._buf for ap in self._input_aps]
        iap_masks = [ap._delay._mask for ap in self._input_aps]
        iap_wps = [ap._delay._write_pos for ap in self._input_aps]
        iap_lens = [ap._length for ap in self._input_aps]
        iap_gs = [ap._g for ap in self._input_aps]

        # Tank modulated allpasses
        tap_bufs = [ap._delay._buf for ap in self._tank_aps]
        tap_masks = [ap._delay._mask for ap in self._tank_aps]
        tap_wps = [ap._delay._write_pos for ap in self._tank_aps]
        tap_lens = [ap._base_length for ap in self._tank_aps]
        tap_gs = [ap._g for ap in self._tank_aps]

        # Tank delays
        td_bufs = [dl._buf for dl in self._tank_delays]
        td_masks = [dl._mask for dl in self._tank_delays]
        td_wps = [dl._write_pos for dl in self._tank_delays]
        td_lens = self._tank_delay_lengths

        # Tank delays 2
        td2_bufs = [dl._buf for dl in self._tank_delays2]
        td2_masks = [dl._mask for dl in self._tank_delays2]
        td2_wps = [dl._write_pos for dl in self._tank_delays2]
        td2_lens = self._tank_delay2_lengths

        # Damping state
        damp0 = self._damp[0]._state
        damp1 = self._damp[1]._state

        # Tank cross-feed state
        ts0 = self._tank_state[0]
        ts1 = self._tank_state[1]

        # LFO
        lfo_phase = self._lfo_phase
        TWO_PI = 2.0 * math.pi

        # Output tap positions (local lists)
        tl = self._taps_l
        tr = self._taps_r

        # --- Main loop: everything inlined ---
        for i in range(n):
            x = float(input_buf[i])
            dry_sample = x

            # Pre-delay: write then read
            pd_buf[pd_wp & pd_mask] = x
            pd_wp += 1
            x = float(pd_buf[(pd_wp - predelay_samps) & pd_mask])

            # Input diffusion: 4 allpass stages inlined
            for j in range(4):
                buf = iap_bufs[j]
                mask = iap_masks[j]
                wp = iap_wps[j]
                dl = iap_lens[j]
                g = iap_gs[j]
                delayed = float(buf[(wp - dl) & mask])
                v = x + g * delayed
                buf[wp & mask] = v
                iap_wps[j] = wp + 1
                x = delayed - g * v

            # LFO (math.sin/cos — ~10x faster than np.sin for scalars)
            lfo1 = math.sin(lfo_phase) * mod_depth
            lfo2 = math.cos(lfo_phase) * mod_depth
            lfo_phase += lfo_inc
            if lfo_phase > TWO_PI:
                lfo_phase -= TWO_PI

            # --- Tank loop 0 ---
            tank_in = x + decay * ts1

            # Modulated allpass 0
            rd = tap_lens[0] + lfo1
            if rd < 1.0:
                rd = 1.0
            rd_int = int(rd)
            rd_frac = rd - rd_int
            b0 = tap_bufs[0]
            m0 = tap_masks[0]
            w0 = tap_wps[0]
            a_val = float(b0[(w0 - rd_int) & m0])
            b_val = float(b0[(w0 - rd_int - 1) & m0])
            delayed = a_val + rd_frac * (b_val - a_val)
            g0 = tap_gs[0]
            v = tank_in + g0 * delayed
            b0[w0 & m0] = v
            tap_wps[0] = w0 + 1
            ap_out = delayed - g0 * v

            # Tank delay 0: write + read
            td_bufs[0][td_wps[0] & td_masks[0]] = ap_out
            td_wps[0] += 1
            d0 = float(td_bufs[0][(td_wps[0] - td_lens[0]) & td_masks[0]])

            # Damping 0
            damp0 = damp_inv * d0 + damping * damp0
            d0 = damp0 * decay

            # Tank delay2 0: write + read
            td2_bufs[0][td2_wps[0] & td2_masks[0]] = d0
            td2_wps[0] += 1
            ts0 = float(td2_bufs[0][(td2_wps[0] - td2_lens[0]) & td2_masks[0]])

            # --- Tank loop 1 ---
            tank_in = x + decay * ts0

            # Modulated allpass 1
            rd = tap_lens[1] + lfo2
            if rd < 1.0:
                rd = 1.0
            rd_int = int(rd)
            rd_frac = rd - rd_int
            b1 = tap_bufs[1]
            m1 = tap_masks[1]
            w1 = tap_wps[1]
            a_val = float(b1[(w1 - rd_int) & m1])
            b_val = float(b1[(w1 - rd_int - 1) & m1])
            delayed = a_val + rd_frac * (b_val - a_val)
            g1 = tap_gs[1]
            v = tank_in + g1 * delayed
            b1[w1 & m1] = v
            tap_wps[1] = w1 + 1
            ap_out = delayed - g1 * v

            # Tank delay 1: write + read
            td_bufs[1][td_wps[1] & td_masks[1]] = ap_out
            td_wps[1] += 1
            d1 = float(td_bufs[1][(td_wps[1] - td_lens[1]) & td_masks[1]])

            # Damping 1
            damp1 = damp_inv * d1 + damping * damp1
            d1 = damp1 * decay

            # Tank delay2 1: write + read
            td2_bufs[1][td2_wps[1] & td2_masks[1]] = d1
            td2_wps[1] += 1
            ts1 = float(td2_bufs[1][(td2_wps[1] - td2_lens[1]) & td2_masks[1]])

            # --- Output taps (inlined reads) ---
            w_td0 = td_wps[0]
            w_td1 = td_wps[1]
            w_td20 = td2_wps[0]
            w_td21 = td2_wps[1]
            m_td0 = td_masks[0]
            m_td1 = td_masks[1]
            m_td20 = td2_masks[0]
            m_td21 = td2_masks[1]

            wet_l = (
                float(td_bufs[0][(w_td0 - tl[0]) & m_td0])
                + float(td_bufs[0][(w_td0 - tl[1]) & m_td0])
                - float(td2_bufs[0][(w_td20 - tl[2]) & m_td20])
                + float(td2_bufs[0][(w_td20 - tl[3]) & m_td20])
                - float(td_bufs[1][(w_td1 - tl[4]) & m_td1])
                - float(td2_bufs[1][(w_td21 - tl[5]) & m_td21])
                - float(td2_bufs[1][(w_td21 - tl[6]) & m_td21])
            )
            wet_r = (
                float(td_bufs[1][(w_td1 - tr[0]) & m_td1])
                + float(td_bufs[1][(w_td1 - tr[1]) & m_td1])
                - float(td2_bufs[1][(w_td21 - tr[2]) & m_td21])
                + float(td2_bufs[1][(w_td21 - tr[3]) & m_td21])
                - float(td_bufs[0][(w_td0 - tr[4]) & m_td0])
                - float(td2_bufs[0][(w_td20 - tr[5]) & m_td20])
                - float(td2_bufs[0][(w_td20 - tr[6]) & m_td20])
            )

            output[i] = dry_sample * dry_gain + (wet_l + wet_r) * 0.25 * wet_gain

        # --- Write back all mutated state ---
        self._predelay._write_pos = pd_wp
        for j in range(4):
            self._input_aps[j]._delay._write_pos = iap_wps[j]
        for j in range(2):
            self._tank_aps[j]._delay._write_pos = tap_wps[j]
            self._tank_delays[j]._write_pos = td_wps[j]
            self._tank_delays2[j]._write_pos = td2_wps[j]
        self._damp[0]._state = damp0
        self._damp[1]._state = damp1
        self._tank_state[0] = ts0
        self._tank_state[1] = ts1
        self._lfo_phase = lfo_phase

        return output

    def clear(self) -> None:
        """Reset all internal state — call when toggling off."""
        self._predelay.clear()
        for ap in self._input_aps:
            ap.clear()
        for ap in self._tank_aps:
            ap.clear()
        for dl in self._tank_delays:
            dl.clear()
        for dl in self._tank_delays2:
            dl.clear()
        for d in self._damp:
            d.clear()
        self._tank_state = [0.0, 0.0]
        self._lfo_phase = 0.0

    def get_state(self) -> dict:
        return {
            "mix": self.mix,
            "decay": self.decay,
            "damping": self.damping,
            "predelay_ms": self.predelay_ms,
            "mod_depth": self.mod_depth,
            "mod_rate": self.mod_rate,
        }

File: native/src/reverb.c
#include "reverb.h"
#include <string.h>
#include <math.h>

/* Dattorro reference lengths at 29761Hz */
#define REF_SR 29761
static const int INPUT_DIFF_LEN[] = {142, 107, 379, 277};
static const float INPUT_DIFF_G[] = {0.75f, 0.75f, 0.625f, 0.625f};
static const int TANK_AP_LEN[] = {672, 908};
static const int TANK_DL_LEN[] = {4453, 4217};
static const int TANK_DL2_LEN[] = {3720, 3163};
static const int TAPS_L[] = {266, 2974, 1913, 1996, 1990, 187, 1066};
static const int TAPS_R[] = {353, 3627, 1228, 2673, 2111, 335, 121};

/* Mode presets: size_mult, mod_depth, mod_rate */
static const float MODES[][3] = {
    {0.30f, 0.3f, 0.5f}, /* ROOM */
    {0.50f, 0.4f, 0.7f}, /* CHAMBER */
    {1.00f, 0.6f, 0.9f}, /* HALL */
    {0.75f, 0.8f, 1.2f}, /* PLATE */
};

static int scale_len(int ref, int sr) { int v = (ref * sr) / REF_SR; return v < 1 ? 1 : v; }
static int next_pow2(int n) { int p = 1; while (p < n) p <<= 1; return p; }

static void dl_init(DelayLine *d, int max_len) {
    int sz = next_pow2(max_len + 16);
    if (sz > REV_MAX_DELAY) sz = REV_MAX_DELAY;
    d->mask = sz - 1; d->wpos = 0;
    memset(d->buf, 0, sizeof(d->buf));
}

static inline void dl_write(DelayLine *d, float s) {
    d->buf[d->wpos & d->mask] = s; d->wpos++;
}

static inline float dl_read(DelayLine *d, int delay) {
    return d->buf[(d->wpos - delay) & d->mask];
}

static inline float dl_read_frac(DelayLine *d, float delay) {
    int di = (int)delay; float frac = delay - di;
    float a = d->buf[(d->wpos - di) & d->mask];
    float b = d->buf[(d->wpos - di - 1) & d->mask];
    return a + frac * (b - a);
}

void reverb_init(Reverb *r) {
    memset(r, 0, sizeof(*r));
    r->mix = 0.3f; r->decay = 0.7f; r->damping = 0.3f;
    r->predelay_ms = 20.0f; r->mod_depth = 0.5f; r->mod_rate = 0.8f;
    r->low_cut_hz = 80.0f; r->mode = 2;

    dl_init(&r->predelay, (int)(0.15f * SR));
    for (int i = 0; i < 4; i++) {
        r->iap_len[i] = scale_len(INPUT_DIFF_LEN[i], SR);
        r->iap_g[i] = INPUT_DIFF_G[i];
        dl_init(&r->iap[i], r->iap_len[i]);
    }
    for (int i = 0; i < 2; i++) {
        r->tap_len[i] = scale_len(TANK_AP_LEN[i], SR);
        r->tap_g[i] = 0.7f;
        dl_init(&r->tap_dl[i], r->tap_len[i] + 64);
        r->td_len[i] = scale_len(TANK_DL_LEN[i], SR);
        dl_init(&r->td[i], r->td_len[i]);
        r->td2_len[i] = scale_len(TANK_DL2_LEN[i], SR);
        dl_init(&r->td2[i], r->td2_len[i]);
    }
}

void reverb_clear(Reverb *r) {
    memset(r->predelay.buf, 0, sizeof(r->predelay.buf));
    for (int i = 0; i < 4; i++) memset(r->iap[i].buf, 0, sizeof(r->iap[i].buf));
    for (int i = 0; i < 2; i++) {
        memset(r->tap_dl[i].buf, 0, sizeof(r->tap_dl[i].buf));
        memset(r->td[i].buf, 0, sizeof(r->td[i].buf));
        memset(r->td2[i].buf, 0, sizeof(r->td2[i].buf));
        r->damp_state[i] = 0; r->tank_state[i] = 0;
    }
    r->lfo_phase = 0; r->hpf_prev_in = 0; r->hpf_prev_out = 0;
}

void reverb_set_mode(Reverb *r, int mode) {
    r->mode = clampf(mode, 0, 3);
    r->mod_depth = MODES[r->mode][1];
    r->mod_rate = MODES[r->mode][2];
}

void reverb_process(Reverb *r, const float *in, float *out, int n) {
    int pd_samps = (int)(r->predelay_ms / 1000.0f * SR);
    if (pd_samps < 1) pd_samps = 1;
    float decay = clampf(r->decay, 0, 0.999f);
    float damp = clampf(r->damping, 0, 0.99f);
    float damp_inv = 1.0f - damp;
    float md = r->mod_depth * 12.0f;
    float lfo_inc = r->mod_rate * TWO_PI / SR;
    float wet = r->mix, dry = 1.0f - r->mix;
    float size_mult = MODES[r->mode][0];

    int td_s[2], td2_s[2];
    int tl[7], tr[7];
    for (int i = 0; i < 2; i++) {
        td_s[i] = (int)(r->td_len[i] * size_mult); if (td_s[i] < 1) td_s[i] = 1;
        td2_s[i] = (int)(r->td2_len[i] * size_mult); if (td2_s[i] < 1) td2_s[i] = 1;
    }
    for (int i = 0; i < 7; i++) {
        tl[i] = (int)(scale_len(TAPS_L[i], SR) * size_mult); if (tl[i] < 1) tl[i] = 1;
        tr[i] = (int)(scale_len(TAPS_R[i], SR) * size_mult); if (tr[i] < 1) tr[i] = 1;
    }

    float lc = clampf(r->low_cut_hz, 20, 500);
    float hpf_rc = 1.0f / (TWO_PI * lc);
    float hpf_dt = 1.0f / SR;
    float hpf_a = hpf_rc / (hpf_rc + hpf_dt);

    for (int i = 0; i < n; i++) {
        float x = in[i], dry_s = x;

        /* Pre-delay */
        dl_write(&r->predelay, x);
        x = dl_read(&r->predelay, pd_samps);

        /* Input diffusion */
        for (int j = 0; j < 4; j++) {
            float d = dl_read(&r->iap[j], r->iap_len[j]);
            float v = x + r->iap_g[j] * d;
            dl_write(&r->iap[j], v);
            x = d - r->iap_g[j] * v;
        }

        float lfo1 = sinf(r->lfo_phase) * md;
        float lfo2 = cosf(r->lfo_phase) * md;
        r->lfo_phase += lfo_inc;
        if (r->lfo_phase > TWO_PI) r->lfo_phase -= TWO_PI;

        /* Tank loop 0 */
        float tin = x + decay * r->tank_state[1];
        float rd = r->tap_len[0] + lfo1; if (rd < 1) rd = 1;
        float delayed = dl_read_frac(&r->tap_dl[0], rd);
        float v = tin + r->tap_g[0] * delayed;
        dl_write(&r->tap_dl[0], v);
        float ap_out = delayed - r->tap_g[0] * v;
        dl_write(&r->td[0], ap_out);
        float d0 = dl_read(&r->td[0], td_s[0]);
        r->damp_state[0] = damp_inv * d0 + damp * r->damp_state[0];
        d0 = r->damp_state[0] * decay;
        dl_write(&r->td2[0], d0);
        r->tank_state[0] = dl_read(&r->td2[0], td2_s[0]);

        /* Tank loop 1 */
        tin = x + decay * r->tank_state[0];
        rd = r->tap_len[1] + lfo2; if (rd < 1) rd = 1;
        delayed = dl_read_frac(&r->tap_dl[1], rd);
        v = tin + r->tap_g[1] * delayed;
        dl_write(&r->tap_dl[1], v);
        ap_out = delayed - r->tap_g[1] * v;
        dl_write(&r->td[1], ap_out);
        float d1 = dl_read(&r->td[1], td_s[1]);
        r->damp_state[1] = damp_inv * d1 + damp * r->damp_state[1];
        d1 = r->damp_state[1] * decay;
        dl_write(&r->td2[1], d1);
        r->tank_state[1] = dl_read(&r->td2[1], td2_s[1]);

        /* Output taps */
        float wl = dl_read(&r->td[0], tl[0]) + dl_read(&r->td[0], tl[1])
                  - dl_read(&r->td2[0], tl[2]) + dl_read(&r->td2[0], tl[3])
                  - dl_read(&r->td[1], tl[4]) - dl_read(&r->td2[1], tl[5])
                  - dl_read(&r->td2[1], tl[6]);
        float wr = dl_read(&r->td[1], tr[0]) + dl_read(&r->td[1], tr[1])
                  - dl_read(&r->td2[1], tr[2]) + dl_read(&r->td2[1], tr[3])
                  - dl_read(&r->td[0], tr[4]) - dl_read(&r->td2[0], tr[5])
                  - dl_read(&r->td2[0], tr[6]);

        float wet_s = (wl + wr) * 0.25f;
        float hpf_out = hpf_a * (r->hpf_prev_out + wet_s - r->hpf_prev_in);
        r->hpf_prev_in = wet_s; r->hpf_prev_out = hpf_out;

        out[i] = dry_s * dry + hpf_out * wet;
    }
}
